web/app/view/MapController.js

```diff
@@ -276,21 +278,62 @@ Ext.define('Traccar.view.MapController', {
     },
 
     selectDevice: function (device, center) {
+        this.hidePopup();
         this.selectMarker(this.latestMarkers[device.get('id')], center);
     },
 
     selectReport: function (position, center) {
         this.selectMarker(this.reportMarkers[position.get('id')], center);
     },
 
+    clickDevice: function (feature) {
+        var record, attributes;
+
+        record = feature.get('record');
+        attributes = record.get('attributes');
+        if (Object.getOwnPropertyNames(attributes).length) {
+            this.showAttributesView(feature);
+        }
+    },
+
     selectFeature: function (feature) {
         var record = feature.get('record');
         if (record) {
             if (record instanceof Traccar.model.Device) {
                 this.fireEvent('selectDevice', record, false);
+                this.fireEvent('clickDevice', feature);
             } else {
                 this.fireEvent('selectReport', record, false);
             }
         }
+    },
+
+    clickMap: function () {
+        this.hidePopup();
+    },
+
+    showAttributesView: function (feature) {
+        var coordinates, popupOverlay, record;
```

Comment:
`coordinates` and `popupOverlay` are only used once. Why do you need separate variables?